File: src/lib/clientApi/status/peer.ts
import {
  type CandidatePairDiagnostics,
  type ConnectionType,
  type PeerDiagnostics,
  type PeerPath,
  type PeerStatus,
} from "../../../types/client";

import { RELAY_PRESENTATION_FRESH_MS } from "../types";

function endpointHost(endpoint: string | null | undefined): string | null {
  const value = endpoint?.trim();
  if (!value) return null;
  if (value.startsWith("[")) {
    const end = value.indexOf("]");
    return end > 1 ? value.slice(1, end).toLowerCase() : null;
  }
  const separator = value.lastIndexOf(":");
  return (separator > 0 ? value.slice(0, separator) : value).toLowerCase();
}

function isPrivateEndpoint(endpoint: string | null | undefined): boolean {
  const host = endpointHost(endpoint);
  if (!host) return false;
  if (host === "localhost" || host === "::1") return true;
  if (host.startsWith("fe80:") || host.startsWith("fc") || host.startsWith("fd")) return true;
  const octets = host.split(".").map(part => Number(part));
  if (octets.length !== 4 || octets.some(part => !Number.isInteger(part))) return false;
  const [a, b] = octets;
  return (
    a === 10 ||
    a === 127 ||
    (a === 169 && b === 254) ||
    (a === 172 && b >= 16 && b <= 31) ||
    (a === 192 && b === 168)
  );
}

function directPairForPresentation(peer: PeerDiagnostics): CandidatePairDiagnostics | null {
  return peer.current_direct_pair ?? peer.selected_pair ?? null;
}

function directPairLatencyMs(peer: PeerDiagnostics): number | null {
  const pair = directPairForPresentation(peer);
  return pair?.rtt_ewma_ms ?? pair?.rtt_ms ?? peer.direct.latency_ms;
}

function minNullable(values: Array<number | null | undefined>): number | null {
  const present = values.filter((value): value is number => value != null);
  return present.length ? Math.min(...present) : null;
}

function pathSuccessAgeMs(peer: PeerDiagnostics, path: PeerPath): number | null {
  if (path === "relay") return peer.relay.last_success_age_ms;
  if (path === "direct") return peer.direct.last_success_age_ms;
  return minNullable([peer.direct.last_success_age_ms, peer.relay.last_success_age_ms]);
}

function hasFreshRelayConfirmation(peer: PeerDiagnostics): boolean {
  return (
    peer.relay.last_success_age_ms != null &&
    peer.relay.last_success_age_ms <= RELAY_PRESENTATION_FRESH_MS &&
    peer.relay.consecutive_failures === 0
  );
}

function controlLastSeenAgeMs(peer: PeerDiagnostics): number | null {
  if (!peer.last_seen) return null;
  const timestampMs = peer.last_seen < 10_000_000_000 ? peer.last_seen * 1000 : peer.last_seen;
  const ageMs = Date.now() - timestampMs;
  return Number.isFinite(ageMs) ? Math.max(0, ageMs) : null;
}

function connectionPresentation(
  peer: PeerDiagnostics,
  path: PeerPath
): { type: ConnectionType; label: string; detail: string } {
  const pair = directPairForPresentation(peer);
  const endpoint = pair?.remote_endpoint ?? peer.endpoint ?? null;
  const reason = peer.current_path_selection?.reason;
  const relayHedged = peer.current_path_selection?.relay_hedged === true;

  if (path === "offline") {
    if (peer.online === false) {
      return {
        type: "offline",
        label: "离线",
        detail: "控制面标记该设备离线；等待对端重新注册或续租",
      };
    }
    return {
      type: "offline",
      label: "不可达",
      detail:
        peer.warning ??
        peer.relay.last_error ??
        peer.direct.last_error ??
        "当前没有可用路径",
    };
  }

  if (path === "connecting") {
    const waitingForRelay =
      peer.state === "fallback_to_relay" || peer.current_path_selection?.path === "relay";
    return {
      type: "connecting",
      label: waitingForRelay ? "中继确认中" : "连接中",
      detail:
        reason ??
        (waitingForRelay
          ? `直连仍在探测，正在等待 relay peer 确认${peer.direct.last_error ? `：${peer.direct.last_error}` : ""}`
          : endpoint
            ? `正在验证 ${endpoint}`
            : "正在建立可用路径"),
    };
  }

  if (path === "relay") {
    const directFailure = peer.direct.last_error ? `；直连不可用：${peer.direct.last_error}` : "";
    return {
      type: "relay",
      label: peer.direct.last_error ? "中继兜底" : "中继",
      detail:
        (reason ?? (peer.relay_server ? `通过 ${peer.relay_server}` : "当前流量经中继转发")) +
        directFailure,
    };
  }

  if (path === "direct_trial") {
    return {
      type: "direct_trial",
      label: "直连试探",
      detail: reason ?? (endpoint ? `正在验证 ${endpoint}` : "正在验证直连路径"),
    };
  }

  if (path === "direct") {
    const directType = pair?.direct_type ?? peer.direct_type;
    if (pair?.is_public_udp_direct || directType === "public_udp") {
      return {
        type: "public_direct",
        label: relayHedged ? "公网直连 + 中继备用" : "公网直连",
        detail: relayHedged
          ? reason ?? (endpoint ? `当前直连端点 ${endpoint}，同时发送中继备用流量` : "当前走公网 UDP 直连，并启用中继备用")
          : endpoint
            ? `当前直连端点 ${endpoint}`
            : "当前走公网 UDP 直连",
      };
    }
    if (directType === "lan") {
      return {
        type: "lan_direct",
        label: "局域网直连",
        detail: endpoint ? `当前直连端点 ${endpoint}` : "当前走局域网直连",
      };
    }
    if (pair?.is_overlay_direct || directType === "overlay") {
      return {
        type: "overlay_direct",
        label: "Overlay 直连",
        detail: endpoint ? `当前直连端点 ${endpoint}` : "当前走 overlay 端点直连",
      };
    }
    if (endpoint && isPrivateEndpoint(endpoint)) {
      return {
        type: "lan_direct",
        label: "局域网直连",
        detail: `当前直连端点 ${endpoint}`,
      };
    }
    if (directType === "probing") {
      return {
        type: "direct_trial",
        label: "直连试探",
        detail: endpoint ? `正在验证 ${endpoint}` : "正在验证直连路径",
      };
    }
    return {
      type: "direct",
      label: relayHedged ? "直连 + 中继备用" : "直连",
      detail: relayHedged
        ? reason ?? (endpoint ? `当前直连端点 ${endpoint}，同时发送中继备用流量` : "当前走直连路径，并启用中继备用")
        : endpoint
          ? `当前直连端点 ${endpoint}`
          : "当前走直连路径",
    };
  }

  return {
    type: "unknown",
    label: "未知",
    detail: "当前路径类型未知",
  };
}

export function mapPeer(peer: PeerDiagnostics): PeerStatus {
  const selection = peer.current_path_selection;
  const isDirectTrial =
    selection?.path === "direct" &&
    selection?.reason_code === "path_direct_trial" &&
    (selection?.relay_hedged === true ||
      peer.active_path === null ||
      peer.active_path === undefined);
  const path: PeerPath = isDirectTrial
    ? "direct_trial"
    : peer.online === false
      ? "offline"
      : peer.active_path ??
        (selection?.path === "relay" && hasFreshRelayConfirmation(peer)
          ? "relay"
          : selection?.path === "relay" ||
              peer.state === "fallback_to_relay" ||
              peer.state === "hole_punching" ||
              peer.state === "connecting"
            ? "connecting"
            : "offline");
  const pathErrors = [peer.direct, peer.relay]
    .filter(health => health.last_error)
    .sort(
      (left, right) =>
        (left.last_failure_age_ms ?? Number.POSITIVE_INFINITY) -
        (right.last_failure_age_ms ?? Number.POSITIVE_INFINITY)
    );
  const lastActiveMs =
    peer.online === false
      ? controlLastSeenAgeMs(peer) ?? pathSuccessAgeMs(peer, path) ?? peer.connected_for_ms
      : pathSuccessAgeMs(peer, path) ?? peer.connected_for_ms ?? controlLastSeenAgeMs(peer);
  const latencyMs =
    path === "direct"
      ? directPairLatencyMs(peer)
      : path === "direct_trial"
        ? directPairLatencyMs(peer) ?? peer.relay.latency_ms
        : path === "relay"
          ? peer.relay.latency_ms
          : null;
  const presentation = connectionPresentation(peer, path);
  return {
    id: peer.node_id,
    name: peer.device_name?.trim() || peer.node_id.slice(0, 12),
    virtualIp: peer.virtual_ip,
    state: peer.state,
    path,
    connectionType: presentation.type,
    connectionLabel: presentation.label,
    connectionDetail: presentation.detail,
    latencyMs,
    endpoint: peer.endpoint ?? "",
    natType: peer.nat_type || "unknown",
    lastActiveMs,
    bytesSent: peer.bytes_sent,
    bytesReceived: peer.bytes_received,
    relayServer: peer.relay_server,
    lastError: pathErrors[0]?.last_error ?? null,
    candidates: peer.candidates,
    directHealth: peer.direct,
    relayHealth: peer.relay,
    selectedPair: peer.selected_pair,
    currentDirectPair: peer.current_direct_pair,
    pathSelectionReason: peer.current_path_selection?.reason ?? null,
    pathSelectionReasonCode: peer.current_path_selection?.reason_code ?? null,
  };
}

File: apps/flutter_client/lib/core/models/diagnostics_models/peer.dart
part of '../diagnostics_models.dart';

// The daemon's default Direct/Relay data-plane keepalive cadence is 25s. A
// verified RTT is displayable for three probe intervals; after that the UI
// must fail closed instead of presenting a historical number as live.
const int _rttFreshnessWindowMs = 3 * 25 * 1000;

class PeerSnapshot {
  PeerSnapshot({
    required this.nodeId,
    required this.deviceName,
    this.platform = '',
    required this.appVersion,
    required this.virtualIp,
    required this.endpoint,
    required this.natType,
    required this.online,
    required this.lastSeen,
    required this.remoteRelayLatencyMs,
    required this.state,
    required this.activePath,
    required this.directType,
    required this.isRelay,
    required this.bytesSent,
    required this.bytesReceived,
    required this.relayServer,
    required this.warning,
    required this.connectedForMs,
    required this.direct,
    required this.relay,
    required this.currentPathSelection,
    required this.relayConfirmedEndpoint,
    required this.relayConfirmedGeneration,
    required this.pathObservability,
  });

  final String nodeId;
  final String deviceName;

  /// Platform reported by newer daemons/control-plane snapshots. Older
  /// snapshots omit it and the UI falls back to a conservative name heuristic.
  final String platform;
  final String appVersion;
  final String virtualIp;
  final String? endpoint;
  final String natType;
  final bool online;
  final int lastSeen;

  /// RTT from the remote peer to that peer's selected relay. This is topology
  /// diagnostics only and must never be rendered as this client's peer RTT.
  final int? remoteRelayLatencyMs;
  final String state;
  final String? activePath;
  final String directType;
  final bool isRelay;
  final int bytesSent;
  final int bytesReceived;
  final String? relayServer;
  final String? warning;
  final int? connectedForMs;
  final PathHealthSnapshot direct;
  final PathHealthSnapshot relay;
  final PathSelectionSnapshot? currentPathSelection;

  /// Relay endpoint whose ingress carried the confirming forced-relay probe
  /// ACK (daemon `relay_confirmed_endpoint`).  Present ONLY after a matching
  /// encrypted relay probe ACK — never from a TCP/TLS connect, a queued
  /// registration, or a candidate RTT.
  final String? relayConfirmedEndpoint;

  /// Network generation of the relay probe ACK confirmation.
  final int? relayConfirmedGeneration;

  final PathObservabilitySnapshot pathObservability;

  factory PeerSnapshot.fromJson(JsonMap json) {
    final selectionJson = _mapOrNull(json['current_path_selection']);
    return PeerSnapshot(
      nodeId: _string(json['node_id']),
      deviceName: _string(json['device_name']),
      platform: _string(
        json['platform'] ?? json['device_platform'] ?? json['os'],
      ),
      appVersion: _string(json['app_version']),
      virtualIp: _string(json['virtual_ip']),
      endpoint: _nullableString(json['endpoint']),
      natType: _string(json['nat_type'], 'unknown'),
      // Lifecycle evidence is fail-closed. A missing field from a partial or
      // mixed-version snapshot must not make a stale peer appear connected.
      online: _bool(json['online'], false),
      lastSeen: _int(json['last_seen']),
      remoteRelayLatencyMs: _intOrNull(json['remote_relay_latency_ms']),
      state: _string(json['state'], 'unknown'),
      activePath: _nullableString(json['active_path']),
      directType: _string(json['direct_type'], 'unknown'),
      isRelay: _bool(json['is_relay']),
      bytesSent: _int(json['bytes_sent']),
      bytesReceived: _int(json['bytes_received']),
      relayServer: _nullableString(json['relay_server']),
      warning: _nullableString(json['warning']),
      connectedForMs: _intOrNull(json['connected_for_ms']),
      direct: PathHealthSnapshot.fromJson(_map(json['direct'])),
      relay: PathHealthSnapshot.fromJson(_map(json['relay'])),
      currentPathSelection: selectionJson == null
          ? null
          : PathSelectionSnapshot.fromJson(selectionJson),
      relayConfirmedEndpoint: _nullableString(json['relay_confirmed_endpoint']),
      relayConfirmedGeneration: _intOrNull(json['relay_confirmed_generation']),
      pathObservability: _pathObservabilityOrEmpty(json['path_observability']),
    );
  }

  String get displayName {
    final trimmed = deviceName.trim();
    if (trimmed.isNotEmpty) return trimmed;
    return nodeId.length <= 12 ? nodeId : nodeId.substring(0, 12);
  }

  String get path {
    if (!online) return 'offline';
    // A daemon state/selection is not a usable path until the corresponding
    // encrypted evidence is present. Fail closed so a relay reconnect or a
    // candidate probe cannot be rendered as an established connection.
    if (activePath == 'relay' && _hasRelayConfirmation) return 'relay';
    if (activePath == 'direct' && state == 'direct') return 'direct';
    // Keep the confirmed relay active while Direct is only a background trial.
    // A selector snapshot may already say `direct` before the encrypted Direct
    // ACK is committed to `active_path`; it must not make the UI hide the
    // verified relay in that interval.
    if (_hasRelayConfirmation &&
        (state == 'relay' || isRelay || activePath == 'relay')) {
      return 'relay';
    }
    final selected = currentPathSelection?.path;
    if (selected == 'direct') {
      return currentPathSelection?.directConfirmed == true
          ? 'direct'
          : 'direct_trial';
    }
    if (selected == 'relay' && _hasRelayConfirmation) return 'relay';
    // The daemon may retain `state == direct` as background encrypted proof
    // while relay-first promotion is still waiting for the same-generation
    // relay ACK.  Without an active path, this is probing—not a usable Direct
    // connection and not a latency sample.
    if (state == 'direct') return 'probing';
    if (state == 'relay' || isRelay || activePath == 'relay') return 'probing';
    if (state == 'fallback_to_relay' ||
        state == 'hole_punching' ||
        state == 'connecting') {
      return 'probing';
    }
    // Roster presence and path usability are separate facts. An online peer
    // with no verified path is still online; it is probing until encrypted
    // Direct/Relay evidence arrives.
    return 'probing';
  }

  String get connectionType {
    if (!online) return 'offline';
    if (path == 'direct') {
      return directType == 'unknown' ? 'direct' : directType;
    }
    if (path == 'relay') return 'relay';
    if (path == 'direct_trial' || path == 'probing') return 'probing';
    return 'probing';
  }

  int? get latencyMs {
    // A latency is only meaningful for a VERIFIED usable path.  A candidate
    // probe's RTT (e.g. the 8ms UDP punch) proves nothing about the data
    // path: it must never be displayed or counted as the connection latency.
    if (!online) return null;
    if (path == 'direct') {
      return isDirectVerified ? direct.displayLatencyMs : null;
    }
    if (path == 'relay') {
      // `remoteRelayLatencyMs` is the remote daemon's RTT to its own relay,
      // not this daemon's end-to-end RTT to the peer.  Only the locally timed
      // and verified relay path sample is a displayable peer latency.
      return isRelayVerified ? relay.displayLatencyMs : null;
    }
    return null;
  }

  /// Candidate-probe RTT (UDP punch / STUN-style measurement).  This is NOT a
  /// connection latency: the peer's data path is not verified until
  /// [path] is `direct` or `relay`.  Used ONLY for the "探测中" label.
  int? get probeLatencyMs {
    if (!online) return null;
    if (path == 'direct_trial') return direct.displayLatencyMs;
    if (path == 'probing' &&
        currentPathSelection?.path == 'direct' &&
        currentPathSelection?.directConfirmed != true) {
      return direct.displayLatencyMs;
    }
    return null;
  }

  /// Whether the relay path to this peer is VERIFIED by a matching encrypted
  /// relay probe ACK (`relay_confirmed_endpoint` — daemon authority).  A
  /// transport connect, a queued registration, or a candidate RTT never sets
  /// this.
  bool get isRelayVerified => path == 'relay' && _hasRelayConfirmation;

  /// Whether the direct path to this peer is VERIFIED by a matching encrypted
  /// direct validation exchange (daemon `active_path == direct` requires the
  /// validation ACK).  Candidate probing never sets this.
  bool get isDirectVerified => path == 'direct' && state == 'direct';

  DateTime? get lastSeenAt {
    if (lastSeen <= 0) return null;
    final timestampMs = lastSeen < 10000000000 ? lastSeen * 1000 : lastSeen;
    return DateTime.fromMillisecondsSinceEpoch(timestampMs);
  }

  int get sortTimestampMs {
    final seen = lastSeenAt?.millisecondsSinceEpoch;
    if (seen != null) return seen;
    if (connectedForMs != null) {
      return DateTime.now().millisecondsSinceEpoch - connectedForMs!;
    }
    return 0;
  }

  bool get _hasRelayConfirmation =>
      relayConfirmedEndpoint != null &&
      relayConfirmedEndpoint!.isNotEmpty &&
      relayConfirmedGeneration != null;

  String? get lastError {
    if (warning case final warning?) return warning;
    return switch (path) {
      'direct' => direct.visibleLastError,
      'relay' => relay.visibleLastError ?? direct.visibleLastError,
      _ => direct.visibleLastError ?? relay.visibleLastError,
    };
  }
}

class PathHealthSnapshot {
  PathHealthSnapshot({
    required this.lastSuccessAgeMs,
    required this.lastFailureAgeMs,
    required this.consecutiveFailures,
    required this.lastError,
    required this.lastErrorCode,
    required this.latencyMs,
    required this.rttEwmaMs,
  });

  final int? lastSuccessAgeMs;
  final int? lastFailureAgeMs;
  final int consecutiveFailures;
  final String? lastError;
  final String? lastErrorCode;
  final int? latencyMs;
  final int? rttEwmaMs;

  /// The UI shows the smoothed RTT for this verified path. The raw sample is
  /// only a compatibility fallback for older daemons. Once the daemon reports
  /// that the last sample is older than the freshness window, fail closed and
  /// render `--` rather than a stale latency.
  ///
  /// This is the encrypted overlay/data-path RTT, not an ICMP echo measurement.
  int? get displayLatencyMs {
    if (lastSuccessAgeMs != null && lastSuccessAgeMs! > _rttFreshnessWindowMs) {
      return null;
    }
    return rttEwmaMs ?? latencyMs;
  }

  String? get visibleLastError {
    if (_isNetworkGenerationRefresh(lastErrorCode, lastError)) return null;
    return lastError;
  }

  factory PathHealthSnapshot.fromJson(JsonMap json) {
    return PathHealthSnapshot(
      lastSuccessAgeMs: _intOrNull(json['last_success_age_ms']),
      lastFailureAgeMs: _intOrNull(json['last_failure_age_ms']),
      consecutiveFailures: _int(json['consecutive_failures']),
      lastError: _nullableString(json['last_error']),
      lastErrorCode: _nullableString(json['last_error_code']),
      latencyMs: _intOrNull(json['latency_ms']),
      rttEwmaMs: _intOrNull(json['rtt_ewma_ms']),
    );
  }
}

bool _isNetworkGenerationRefresh(String? code, String? reason) {
  const networkGenerationChanged = 'network_generation_changed';
  final normalizedCode = code?.trim();
  final normalizedReason = reason?.trim();
  return normalizedCode == networkGenerationChanged ||
      normalizedReason == networkGenerationChanged ||
      normalizedReason?.startsWith('$networkGenerationChanged:') == true;
}

class PathSelectionSnapshot {
  PathSelectionSnapshot({
    required this.path,
    required this.reasonCode,
    required this.reason,
    required this.directConfirmed,
    required this.relayHedged,
  });

  final String? path;
  final String reasonCode;
  final String reason;
  final bool directConfirmed;
  final bool relayHedged;

  factory PathSelectionSnapshot.fromJson(JsonMap json) {
    return PathSelectionSnapshot(
      path: _nullableString(json['path']),
      reasonCode: _string(json['reason_code']),
      reason: _string(json['reason']),
      directConfirmed: _bool(json['direct_confirmed']),
      relayHedged: _bool(json['relay_hedged']),
    );
  }
}
